src/loaders/anndata-loaders/MatrixZarrLoader.js

```diff
@@ -31,25 +31,29 @@ const concatenateColumnVectors = (arr) => {
 /**
  * Loader for converting zarr into the a cell x gene matrix for use in Genes/Heatmap components.
  */
-export default class MatrixZarrLoader extends AbstractTwoStepLoader {
+export default class MatrixZarrLoader extends AnnDataLoader {
   /**
-   * Class method for loading the genes list from AnnData.var.
+   * Class method for loading the genes list from AnnData.var,
+   * filtered if a there is a `geneFilterZarr` present in the view config.
    * @returns {Promise} A promise for the zarr array contianing the gene names.
    */
-  async loadGeneNames() {
-    if (this.geneNames) {
-      return this.geneNames;
+  async loadFilteredGeneNames() {
+    if (this.filteredGeneNames) {
+      return this.filteredGeneNames;
     }
     const { geneFilter: geneFilterZarr } = this.options;
     let geneFilter;
     if (geneFilterZarr) {
       geneFilter = await this.dataSource.getFlatArrDecompressed(geneFilterZarr);
     }
-    const { _index } = await this.dataSource.getJson('var/.zattrs');
-    if (this.geneNames) {
-      return this.geneNames;
+    const geneNames = this.loadGeneNames();
+    if (this.filteredGeneNames) {
+      return this.filteredGeneNames;
```

Comment:
why is this conditional block here and above? I don't see any assignment to `this.filteredGeneNames` from line `44-50`?

Reply:
It's probably overkill, but the fetching/decompressing of the filter can take time and this object is shared, so checking the cache again could save re-assigning `this.filteredGeneNames` again

Reply:
Hmm I see, but I'm not convinced there is an issue. Users of the class await promises on `getFlatArrayDecompressed` and `loadGeneNames`. The _first_ caller of either of those methods creates the shared promises (in the `DataSource`) for other callers. Once those data are resolved,  the only effect this method has is applying the filter to the resolved data which is super inexpensive.

Just realized we can probably await these promises concurrently (and avoid iterating over the data if there isn't a gene filter).


```javascript
const getFilterFn = async () => {
   if (!geneFilter) return (data) => data;
   const geneFilter = this.dataSource.getFlatArrDecompressed(geneFilterZarr);
   return (data) => data.filter((_, j) =>  geneFilter[j]);
}

this.filteredGeneNames = Promise
  .all([this.loadGeneNames(), getFilterFn()])
  .then(([data, filter]) => filter(data));
```

Reply:
Cool, I like that!

Reply:
We got rid of the `loadGeneNames` method, but other than that, it works great.